Revise the contents of the storage abstraction doc (#1119)
scalar-labs/scalardb: docs/storage-abstraction.md

> # Storage Abstraction

Author: Since many parts of this doc explain the Storage API, should we rename the title to something like the following?:

```suggestion
# Storage Abstraction and API Guide
```

Reviewer: Agreed!

- If you don't need transactions for some of the data in your application, you can use the Storage API to partially avoid transactions, which results in faster execution.

{% capture notice--warning %}
**Attention**

Directly using the Storage API or mixing the Transactional API and the Storage API could cause unexpected behavior. For example, since the Storage API cannot provide transaction capability, the API could cause anomalies or data inconsistency if failures occur when executing operations.

Therefore, you should be *very* careful about using the Storage API and use it only if you know exactly what you are doing.
{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>

## Storage API Example

This section explains how the Storage API can be used in a basic electronic money application.

{% capture notice--warning %}
**Attention**

The electronic money application is simplified for this example and isn’t suitable for a production environment.
{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>

### ScalarDB configuration

> Before you should configure ScalarDB in the same way mentioned in [Getting Started with ScalarDB](getting-started-with-scalardb.md).

Reviewer 2: I feel there seems something wrong in the sentence. (e.g., `Before` is needed?)

Author: Ah, the beginning of this sentence is definitely missing something. How does this look?:

```suggestion
Before you begin, you should configure ScalarDB in the same way mentioned in [Getting Started with ScalarDB](getting-started-with-scalardb.md).
```

}
```

> To apply the schema, go to the [`scalardb Releases`](https://github.com/scalar-labs/scalardb/releases) page and download the ScalarDB Schema Loader that matches the version of ScalarDB that you are using to the `getting-started` folder.

Reviewer 3: ```suggestion
To apply the schema, go to the [ScalarDB Releases](https://github.com/scalar-labs/scalardb/releases) page and download the ScalarDB Schema Loader that matches the version of ScalarDB that you are using to the `getting-started` folder.
```

Author: I think we have a similar sentence in other docs, so I'll update those to match this style later.

> Then, run the following command, replacing `<VERSION>` with the version of the ScalarDB Schema Loader that you downloaded:

Reviewer 3: > replacing `<VERSION>` with

[minor] The following command uses lower-case `<version>`. We should use the same style?

Author: Good catch! I've changed `<version>` to `<VERSION>`.

```console
$ java -jar scalardb-schema-loader-<VERSION>.jar --config scalardb.properties -f emoney-storage.json
```

### Example code

After loading the schema, you can execute transactions and retrieve data in the basic electronic money application. For details about which types of transactions are supported in the application and how to use the application, see [Execute transactions and retrieve data in the basic electronic money application](getting-started-with-scalardb.md#execute-transactions-and-retrieve-data-in-the-basic-electronic-money-application).

{% capture notice--warning %}
**Attention**

As previously mentioned, since the Storage API cannot provide transaction capability, the API could cause anomalies or data inconsistency if failures occur when executing operations. Therefore, you should be *very* careful about using the Storage API and use it only if you know exactly what you are doing.
{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>

### Reference

> To see the source code for the electronic money application used in this tutorial, see [`ElectronicMoney.java`](https://scalardb.scalar-labs.com/docs/latest/getting-started/src/main/java/sample/ElectronicMoney.java).

Reviewer 4: This code isn't for the Storage API, but the Transaction API. I don't think we have a separate source code file for this. So, I think we need to restore the Java code.

Author: Understood. I've re-added the example code here.

}
```

To apply the schema, go to the [ScalarDB Releases](https://github.com/scalar-labs/scalardb/releases) page and download the ScalarDB Schema Loader that matches the version of ScalarDB that you are using to the `getting-started` folder.

Then, run the following command, replacing `<VERSION>` with the version of the ScalarDB Schema Loader that you downloaded:

```console
$ java -jar scalardb-schema-loader-<VERSION>.jar --config scalardb.properties -f emoney-storage.json
```

### Example code

The following is the source code for the electronic money application that uses the Storage API.

> To use this source code, create a file named `ElectronicMoneyStorage.java` in the `scalardb/docs/getting-started/src/main/java/sample` directory. Then, add the following Java code to the file.

Author: @brfrn169 After re-adding the Java source code to this doc, I added these sentences since I felt like we should guide users on what to do with the source code. What do you think about this part?

Reviewer 4: @josh-wong I'm not sure just copying the source code works. Did you test is and check if it works?

Author: @brfrn169 Sorry, I made an assumption here since I didn't see instructions on what users should do with this source code.

When I try the method I'm seeing errors when I try running this code in the way that I described. I've also tried copying this code to `ElectronicMoney.java` and pasting it over the existing code, but I still encounter errors.

Could you please tell me where users should apply this source code to so that we can provide some guidance?

Reviewer 4: @josh-wong How about we just place the example source code here and skip the detailed run instructions because they might get a bit tricky? Given this document is for experts, I feel it should be fine. What do you think?

Author: @brfrn169 I'm OK with that for now. I've removed the instructions that I previously added and changed this part to be informational as it was before. We can update this section later based on feedback, if necessary.

You can create a secondary index as follows:

```java
// Create a secondary index on column "c5" for table "ns.tbl". If a secondary index already exists, an exception will be thrown.
admin.createIndex("ns", "tbl", "c5");

// Create the secondary index only if it does not already exist.
boolean ifNotExists = true;
admin.createIndex("ns", "tbl", "c5", ifNotExists);

// Create the secondary index with options.
Map<String, String> options = ...;
admin.createIndex("ns", "tbl", "c5", options);
```

> ### Add a new column to a table

Author: Got it, thanks! Originally, this section was a level 3 heading, so I assumed the subsequent headings were supposed to be level 3 headings as well. I've applied the proper level of headings here and throughout.

You can add a new, non-partition key column to a table as follows:

```java
// Add a new column "c6" with the INT data type to the table "ns.tbl".
admin.addNewColumnToTable("ns", "tbl", "c6", DataType.INT)
```

{% capture notice--warning %}
**Attention**

You should carefully consider adding a new column to a table because the execution time may vary greatly depending on the underlying storage. Please plan accordingly and consider the following, especially if the database runs in production:

- **For Cosmos DB for NoSQL and DynamoDB:** Adding a column is almost instantaneous as the table schema is not modified. Only the table metadata stored in a separate table is updated.
- **For Cassandra:** Adding a column will only update the schema metadata and will not modify the existing schema records. The cluster topology is the main factor for the execution time. Changes to the schema metadata are shared to each cluster node via a gossip protocol. Because of this, the larger the cluster, the longer it will take for all nodes to be updated.

> - **For relational databases (MySQL, Oracle, etc.):** Adding a column may take a very long time to execute. In addition, a table lock may occur.

Reviewer: As I mentioned in the API guide page, it wouldn't take so long for relational databases to add columns. This should also be revised.

Author: Good catch! I've revised this sentence.

{% endcapture %}

<div class="notice--warning">{{ notice--warning | markdownify }}</div>

> ### Truncate a table

Reviewer 4: Ditto.
```suggestion
#### Truncate a table
```

You can truncate a table as follows:

```java
// Truncate the table "ns.tbl".
admin.truncateTable("ns", "tbl");
```

### Drop a secondary index

You can drop a secondary index as follows:

```java
// Drop the secondary index on column "c5" from table "ns.tbl". If the secondary index does not exist, an exception will be thrown.
admin.dropIndex("ns", "tbl", "c5");

// Drop the secondary index only if it exists.
boolean ifExists = true;
admin.dropIndex("ns", "tbl", "c5", ifExists);
```

> ### Drop a table

Reviewer 4: Ditto.
```suggestion
#### Drop a table
```

You can drop a table as follows:

```java
// Drop the table "ns.tbl". If the table does not exist, an exception will be thrown.
admin.dropTable("ns", "tbl");

// Drop the table only if it exists.
boolean ifExists = true;
admin.dropTable("ns", "tbl", ifExists);
```

> ### Drop a namespace

Reviewer 4: Ditto.
```suggestion
#### Drop a namespace
```

You can drop a namespace as follows:

```java
// Drop the namespace "ns". If the namespace does not exist, an exception will be thrown.
admin.dropNamespace("ns");

// Drop the namespace only if it exists.
boolean ifExists = true;
admin.dropNamespace("ns", ifExists);
```

> ### Get existing namespaces

Author: @Torch3333 I noticed that you added this part in PR #989. Just to confirm, this section should only exist in the `master` branch, right? If so, I will remove this section when we backport it to `3.10` and earlier branch versions.

Reviewer 5: Yes, exactly. This section should only exist in the `master` branch.

You can get the tables of a namespace as follows:

```java
// Get the tables of the namespace "ns".
Set<String> tables = admin.getNamespaceTableNames("ns");
```

> ### Get table metadata

Author: @Torch3333 Same as with the `### Get existing namespaces` section, this section should only exist in the `master` branch, right? If so, I will remove this section when we backport it to `3.10` and earlier branch versions.

        .limit(10)
        .build();

// Execute the `Scan` operation.
Scanner scanner = storage.scan(scan);
```

{% capture notice--info %}
**Note**

You can't specify clustering-key boundaries and orderings in `Scan` without specifying a partition key.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

> #### `Put` operation

Author: The structure for the following four sections in this doc are as follows:

```markdown
...
Put operation
  Conditions for Put
Delete operation
  Conditions for Delete
...
```

However, should this structure match the same structure as in `api-guide.md`, as shown below? Or do these sections need to be kept separate in this doc? 

```markdown
...
Put operation
Delete operation
Put and Delete with a condition
  Conditions for Put
  Conditions for Delete
...
```

Reviewer 5: We can use the same structure as in the `api-guide.md`

Author: Got it. Thanks!

> `Put` is an operation to put a record specified by a primary key. The operation behaves as an upsert operation for a record, in which the operation updates the record if the record exists or inserts the record if the record does not exist.

Author: In PR #1084 that contains revisions for the [`api-guide.md`](https://github.com/scalar-labs/scalardb/blob/6f221b039243262a061aabd36f7652a17fcb5f75/docs/api-guide.md?plain=1#L657), we have a note after this paragraph. Do we need that same note here?: 

```suggestion
`Put` is an operation to put a record specified by a primary key. The operation behaves as an upsert operation for a record, in which the operation updates the record if the record exists or inserts the record if the record does not exist.

{% capture notice--info %}
**Note**

When you update an existing record, you need to read the record by using `Get` or `Scan` before using a `Put` operation.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>
```

Reviewer 4: No, we don't need the note for the storage API. It's only for the transaction API.

        .build();
```

#### `Delete` operation

`Delete` is an operation to delete a record specified by a primary key.

You need to create a `Delete` object first, and then you can execute the object by using the `storage.delete()` method as follows:

```java
// Create a `Delete` operation.
Key partitionKey = Key.ofInt("c1", 10);
Key clusteringKey = Key.of("c2", "aaa", "c3", 100L);

Delete delete =
    Delete.newBuilder()
        .namespace("ns")
        .table("tbl")
        .partitionKey(partitionKey)
        .clusteringKey(clusteringKey)
        .build();

// Execute the `Delete` operation.
storage.delete(delete);
```

#### `Put` and `Delete` with a condition

> You can write arbitrary conditions (for example, a bank account balance must be equal to or more than zero) that you require a transaction to meet before being committed by implementing logic that checks the conditions in the transaction. Alternatively, you can write simple conditions in a mutation operation, such as `Put` and `Delete`.
> 
> When a `Put` or `Delete` operation includes a condition, the operation is executed only if the specified condition is met. If the condition is not met when the operation is executed, an exception called `UnsatisfiedConditionException` will be thrown.

Reviewer 5: I rewrote a bit this section since in the case of the storage abstraction, a `NoMutationException` is thrown instead when the condition is not satisfied.
What do you think?
 
```suggestion
You can write arbitrary conditions (for example, a bank account balance must be equal to or more than zero) that you require an operation to meet before being executed by implementing logic that checks the conditions. Alternatively, you can write simple conditions in a mutation operation, such as `Put` and `Delete`.

When a `Put` or `Delete` operation includes a condition, the operation is executed only if the specified condition is met. If the condition is not met when the operation is executed, an exception called `NoMutationException` will be thrown.
```

Author: LGTM! Thank you for revising this part!🙏

        .build();
```

> ##### Conditions for `Put`

Author: @Torch3333 Semi-related to [my previous question that you answered](https://github.com/scalar-labs/scalardb/pull/1119#discussion_r1339815316), I added the following paragraphs here since a description didn't exist. This is the same description from the [api-guide.md](https://github.com/scalar-labs/scalardb/blob/6f221b039243262a061aabd36f7652a17fcb5f75/docs/api-guide.md?plain=1#L731). Does this look OK? 

 I'm mostly wondering if the last sentence is correct since we don't mention `UnsatisfiedConditionException` elsewhere in this document.

Reviewer 5: Thanks. We need to change the content since the behavior differs slightly for the storage abstraction.
Please refer to my other comment.

        .clusteringKey(clusteringKeyForDelete)
        .build();

// Execute the operations.
storage.mutate(Arrays.asList(put, delete));
```

{% capture notice--info %}
**Note**

A Mutate operation only accepts mutations for a single partition; otherwise, an exception will be thrown.

In addition, if you specify multiple conditions in a Mutate operation, the operation will be executed only when all the conditions match.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

> #### Default namespace for CRUD operations

Reviewer 4: I assume this PR should be applied to versions 3.5 and later, correct? Also, as you know, this namespace config was introduced in version 3.9 and later. So, please remove it when applying to versions 3.8 and earlier.

Author: Got it. I've noted this and will remove it from 3.8 and earlier versions.
